### PR TITLE
Check that the env is evar free before using the kernel in Check/Eval

### LINKED ISSUE #17298: Anomaly "File "kernel/constr.ml", line 1482, characters 28-34: Assertion failed."
#### Description of the problem

The code
<details>
  <summary>Reproducible example</summary>

```coq
(* -*- mode: coq; coq-prog-args: ("-emacs" "-w" "-deprecated-native-compiler-option,-native-compiler-disabled" "-R" "." "Top" "-top" "Top.x" "-native-compiler" "ondemand") -*- *)
(* File reduced by coq-bug-minimizer from original input, then from 1913 lines to 227 lines, then from 232 lines to 228 lines *)
(* coqc version 8.18+alpha compiled with OCaml 4.14.0
   coqtop version notebook:/home/user/.opam/default/.opam-switch/build/coq-core.dev/_build/default,master (9cba9b3754e397b55f026340194afb667cf919b6)
   Expected coqc runtime on this file: 0.239 sec *)
Require Coq.btauto.Btauto.

Import List.ListNotations.
Import Coq.NArith.BinNat.

#[global] Obligation Tactic := idtac.

Open Scope list_scope.
Open Scope bool_scope.

Notation binnat := Coq.Numbers.BinNums.N.

Definition normalize_eq_binnat n m (H : n = m) :=
  match N.eq_dec n m with
  | left H1 => H1
  | right H1 => False_rect _ (H1 H)
  end.

Infix "^^" := xorb (at level 40, left associativity) : bool_scope.

Definition length_bin {A} (l : list A) := N.of_nat (length l).

Definition list_nth_bin {A} i l (default : A) := List.nth (N.to_nat i) l default.

Fixpoint list_forall {A} f (l : list A) :=
  match l with
  | [] => True
  | x :: l' => f x /\ list_forall f l'
  end.

Fixpoint list_forall_i_bin_aux {A} f i (l : list A) :=
  match l with
  | [] => True
  | x :: l' => f i x /\ list_forall_i_bin_aux f (N.succ i) l'
  end.

Definition list_forall_i_bin {A} f (l : list A) := list_forall_i_bin_aux f 0%N l.

Fixpoint list_init_aux {A} f s n : list A :=
  match n with
  | 0 => []
  | S n' => f s :: list_init_aux f (S s) n'
  end.

Definition list_init {A} f n : list A := list_init_aux f 0 n.

Definition list_init_bin {A} f n : list A :=
  list_init (fun i => f (N.of_nat i)) (N.to_nat n).

Definition list_seq_bin s n :=
  List.map N.of_nat (List.seq (N.to_nat s) (N.to_nat n)).

Lemma length_bin_list_seq_bin :
  forall s n,
  length_bin (list_seq_bin s n) = n.
Admitted.

Definition list_select_bin {A} values indices (default : A) :=
  List.map (fun i => list_nth_bin i values default) indices.

Record vector {A n} := {
  vector_list : list A;
  vector_wf : length_bin vector_list = n;
}.

Arguments vector : clear implicits.

Notation bitvec := (vector bool).

Definition mk_vector {A n} v_list v_wf : vector A n := {|
  vector_list := v_list;
  vector_wf := normalize_eq_binnat _ _ v_wf;
|}.
Definition vector_normalize {A n} (v : vector A n) : vector A n.
Admitted.

Inductive reference :=
  | reference_Zero
  | reference_One
  | reference_Input (i : binnat)
  | reference_Wire (i : binnat).

Inductive binding :=
  | binding_Immediate (r : reference)
  | binding_Not (r : reference)
  | binding_And (r1 r2 : reference)
  | binding_Or (r1 r2 : reference)
  | binding_Xor (r1 r2 : reference)
  | binding_If (r1 r2 r3 : reference).

Record circuit := {
  circuit_input_count : binnat;
  circuit_wire_count : binnat;
  circuit_wires : list binding;
  circuit_output_wires : list binnat;
}.

Definition reference_wf input_count wire_count r :=
  match r with
  | reference_Zero => True
  | reference_One => True
  | reference_Input i => (i < input_count)%N
  | reference_Wire i => (i < wire_count)%N
  end.

Definition binding_wf input_count wire_count b :=
  match b with
  | binding_Immediate r => reference_wf input_count wire_count r
  | binding_Not r => reference_wf input_count wire_count r
  | binding_And r1 r2 =>
    reference_wf input_count wire_count r1 /\
    reference_wf input_count wire_count r2
  | binding_Or r1 r2 =>
    reference_wf input_count wire_count r1 /\
    reference_wf input_count wire_count r2
  | binding_Xor r1 r2 =>
    reference_wf input_count wire_count r1 /\
    reference_wf input_count wire_count r2
  | binding_If r1 r2 r3 =>
    reference_wf input_count wire_count r1 /\
    reference_wf input_count wire_count r2 /\
    reference_wf input_count wire_count r3
  end.

Definition circuit_wire_count_wf c :=
  length_bin (circuit_wires c) = circuit_wire_count c.

Definition circuit_wires_wf c :=
  list_forall_i_bin (binding_wf (circuit_input_count c)) (circuit_wires c).

Definition circuit_output_wires_wf c :=
  list_forall (fun i => (i < circuit_wire_count c)%N) (circuit_output_wires c).

Definition circuit_wf c :=
  circuit_wire_count_wf c /\ circuit_wires_wf c /\ circuit_output_wires_wf c.

Record circuit_with_wf := {
  circuit_with_wf_circuit : circuit;
  circuit_with_wf_circuit_wf : circuit_wf circuit_with_wf_circuit;
}.

Record circuit_with_wf_and_spec := {
  circuit_with_wf_and_spec_circuit_with_wf : circuit_with_wf;
  circuit_with_wf_and_spec_spec_statement : Prop;
  circuit_with_wf_and_spec_spec : circuit_with_wf_and_spec_spec_statement;
}.

Definition reference_compute inputs intermediates r :=
  match r with
  | reference_Zero => false
  | reference_One => true
  | reference_Input i => list_nth_bin i inputs false
  | reference_Wire i => list_nth_bin i intermediates false
  end.

Definition binding_compute inputs intermediates b :=
  match b with
  | binding_Immediate r => reference_compute inputs intermediates r
  | binding_Not r => negb (reference_compute inputs intermediates r)
  | binding_And r1 r2 =>
    reference_compute inputs intermediates r1 && reference_compute inputs intermediates r2
  | binding_Or r1 r2 =>
    reference_compute inputs intermediates r1 || reference_compute inputs intermediates r2
  | binding_Xor r1 r2 =>
    reference_compute inputs intermediates r1 ^^ reference_compute inputs intermediates r2
  | binding_If r1 r2 r3 =>
    if reference_compute inputs intermediates r1
    then reference_compute inputs intermediates r3
    else reference_compute inputs intermediates r2
  end.

Definition circuit_compute_wires_aux start_intermediates wires inputs :=
  List.fold_left (fun intermediates b =>
    intermediates ++ [binding_compute inputs (start_intermediates ++ intermediates) b]
  ) wires [].

Definition circuit_compute_wires c inputs :=
  circuit_compute_wires_aux [] (circuit_wires c) inputs.

Definition circuit_compute c inputs :=
  list_select_bin (circuit_compute_wires c inputs) (circuit_output_wires c) false.

#[program] Definition circuit_with_wf_compute_vec
  c_with_wf (inputs : bitvec (circuit_input_count (circuit_with_wf_circuit c_with_wf)))
  : bitvec (length_bin (circuit_output_wires (circuit_with_wf_circuit c_with_wf))) :=
  mk_vector (circuit_compute (circuit_with_wf_circuit c_with_wf) (vector_list inputs)) _.
Admit Obligations.

Definition circuit_id input_count := {|
  circuit_input_count := input_count;
  circuit_wire_count := input_count;
  circuit_wires := list_init_bin (fun i => binding_Immediate (reference_Input i)) input_count;
  circuit_output_wires := list_seq_bin 0 input_count;
|}.

Theorem circuit_id_wf :
  forall input_count,
  circuit_wf (circuit_id input_count).
Admitted.

#[program] Definition circuit_id_with_wf_and_spec input_count : circuit_with_wf_and_spec :=
  let c_res_with_wf := {|
    circuit_with_wf_circuit := circuit_id input_count;
    circuit_with_wf_circuit_wf := circuit_id_wf input_count;
  |} in
  {|
    circuit_with_wf_and_spec_circuit_with_wf := c_res_with_wf;
    circuit_with_wf_and_spec_spec_statement :=
      forall (inputs : bitvec input_count),
      circuit_with_wf_compute_vec c_res_with_wf inputs = vector_normalize inputs;
  |}.
Next Obligation.
  intros input_count c_res_with_wf inputs.
apply normalize_eq_binnat.
simpl.
symmetry.
  apply length_bin_list_seq_bin.
Defined.
Next Obligation.
  intros input_count c_res_with_wf inputs.
unfold circuit_with_wf_compute_vec.
simpl.
  Eval simpl in (mk_vector (circuit_compute (circuit_id input_count) (vector_list inputs))
  (circuit_with_wf_compute_vec_obligation_1 c_res_with_wf inputs) =
eq_rect input_count (fun n : binnat => bitvec n) (vector_normalize inputs) (length_bin (list_seq_bin 0 input_count))
  (circuit_id_with_wf_and_spec_obligation_1 input_count)).
```
</details>

gives the following error on a recent master:

```
Welcome to Coq notebook:/home/user/.opam/default/.opam-switch/build/coq-core.dev/_build/default,master (9cba9b3754e397b55f026340194afb667cf919b6)
File "/tmp/tmp.BLzJaQUe2w/o.v", line 228, characters 2-334:
Error:
Anomaly
"File "kernel/constr.ml", line 1482, characters 28-34: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
Raised at Constr.default_evar_handler.(fun) in file "kernel/constr.ml", line 1482, characters 28-40
Called from CClosure.is_irrelevant in file "kernel/cClosure.ml", line 482, characters 42-80
Called from CClosure.shortcut_irrelevant in file "kernel/cClosure.ml", line 486, characters 5-25
Called from CClosure.ref_value_cache in file "kernel/cClosure.ml", line 523, characters 21-80
Called from CClosure.unfold_ref_with_args in file "kernel/cClosure.ml", line 1756, characters 8-42
Called from Reduction.eqappr in file "kernel/reduction.ml", line 469, characters 17-72
Called from CArray.fold_right2.fold in file "clib/cArray.ml", line 242, characters 11-40
Called from Reduction.gen_conv in file "kernel/reduction.ml", line 962, characters 14-91
Called from Reduction.default_conv in file "kernel/reduction.ml" (inlined), line 975, characters 4-28
Called from Typeops.conv_leq in file "kernel/typeops.ml" (inlined), line 30, characters 23-49
Called from Typeops.type_of_apply.apply_rec in file "kernel/typeops.ml", line 251, characters 20-40
Called from Typeops.execute in file "kernel/typeops.ml", line 673, characters 14-49
Called from Typeops.execute_array.(fun) in file "kernel/typeops.ml", line 841, characters 53-66
Called from CArray.Smart.map_i in file "clib/cArray.ml", line 518, characters 15-21
Called from Typeops.execute_array in file "kernel/typeops.ml", line 841, characters 11-90
Called from Typeops.execute in file "kernel/typeops.ml", line 663, characters 25-47
Called from Typeops.execute_array.(fun) in file "kernel/typeops.ml", line 841, characters 53-66
Called from CArray.Smart.map_i in file "clib/cArray.ml", line 518, characters 15-21
Called from Typeops.execute_array in file "kernel/typeops.ml", line 841, characters 11-90
Called from Typeops.execute in file "kernel/typeops.ml", line 663, characters 25-47
Called from Typeops.infer in file "kernel/typeops.ml", line 854, characters 18-36
Called from Arguments_renaming.rename_typing in file "pretyping/arguments_renaming.ml", line 110, characters 10-29
Called from Vernacentries.check_may_eval in file "vernac/vernacentries.ml", line 2062, characters 44-84
Called from Vernacentries.translate_vernac.(fun) in file "vernac/vernacentries.ml", line 2611, characters 8-43
Called from Vernacextend.vtreadproofopt.(fun) in file "vernac/vernacextend.ml", line 159, characters 29-44
Called from Vernacinterp.interp_typed_vernac in file "vernac/vernacinterp.ml", line 20, characters 20-113
Called from Vernacinterp.interp_control.(fun) in file "vernac/vernacinterp.ml", line 221, characters 24-113
Called from Flags.with_modified_ref in file "lib/flags.ml", line 17, characters 14-17
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Vernacinterp.interp_gen.(fun) in file "vernac/vernacinterp.ml", line 274, characters 18-43
Called from Vernacinterp.interp_gen in file "vernac/vernacinterp.ml", line 272, characters 6-279
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Stm.Reach.known_state.reach.(fun) in file "stm/stm.ml", line 2178, characters 16-43
Called from Stm.State.define in file "stm/stm.ml", line 965, characters 6-10
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Stm.Reach.known_state.reach in file "stm/stm.ml", line 2320, characters 4-105
Called from Stm.observe in file "stm/stm.ml", line 2417, characters 4-60
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Vernac.interp_vernac in file "toplevel/vernac.ml", line 68, characters 29-50
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Flags.with_modified_ref in file "lib/flags.ml", line 17, characters 14-17
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Flags.silently in file "lib/flags.ml" (inlined), line 64, characters 19-40
Called from Vernac.load_vernac_core.loop in file "toplevel/vernac.ml", line 120, characters 8-69
Called from Vernac.load_vernac_core in file "toplevel/vernac.ml", line 124, characters 6-19
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Vernac.load_vernac in file "toplevel/vernac.ml", line 180, characters 30-94
Called from Stdlib__List.fold_left in file "list.ml", line 121, characters 24-34
Called from Coqtop.init_toplevel in file "toplevel/coqtop.ml", line 66, characters 20-58
Called from Coqtop.start_coq in file "toplevel/coqtop.ml", line 73, characters 8-28
```

#### Coq Version

```
user@notebook:~$ coqtop -v
The Coq Proof Assistant, version 8.18+alpha
compiled with OCaml 4.14.0
```

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #17298

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot run full ci

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -1829,7 +1829,10 @@ let check_may_eval env sigma redexp rc =
   let uctx = Evd.universe_context_set sigma in
   let env = Environ.push_context_set uctx (Evarutil.nf_env_evar sigma env) in
   let { Environ.uj_val=c; uj_type=ty; } =
-    if Evarutil.has_undefined_evars sigma c then
+    if Evarutil.has_undefined_evars sigma c
+    || List.exists (Context.Named.Declaration.exists (Evarutil.has_undefined_evars sigma))
```
Comment (ppedrot) [inline vernac/vernacentries.ml:1833]: You should use Evarutil.is_ground_env instead.

Comment (SkySkimmer) [reply vernac/vernacentries.ml:1833]: That looks wrong, it doesn't check that the types are evar free
~~~ocaml
let is_ground_env evd env =
  let is_ground_rel_decl = function
    | RelDecl.LocalDef (_,b,_) -> is_ground_term evd (EConstr.of_constr b)
    | _ -> true in
  let is_ground_named_decl = function
    | NamedDecl.LocalDef (_,b,_) -> is_ground_term evd (EConstr.of_constr b)
    | _ -> true in
  List.for_all is_ground_rel_decl (rel_context env) &&
  List.for_all is_ground_named_decl (named_context env)
~~~

Comment (ppedrot) [reply vernac/vernacentries.ml:1833]: Hmm, right. Now I'm a bit surprised that this function is written that way.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**